[18-30] Replace typedefs with alias

AlisdairM opened the pull request:
This set of changes replace (almost) all use of 'typedef' in the library with a type alias instead:
   typedef Original NewName;   ->    using NewName = Original;

Attention was paid to retaining table-like formatting where present, which is worth reviewing in case I made idiosyncratic choices.

Clause 20 has NOT been applied, as that clause is churning a bit at the moment, and I don't want a conflict there to hold up a this review.  The most likely conflict is my own pull request:
   https://github.com/cplusplus/draft/pull/683
I will happily rebase and complete the work once that branch lands (or is deferred again until this branch completes)

A second set of changes that should follow are to update the text that refers to typedefs to perhaps use a term more consistent with the type alias syntax applied here - but I defer to the project editors to suggest such phrasing.

AlisdairM commented:
This pull request addresses https://github.com/cplusplus/draft/issues/581

AlisdairM commented:
Note that I can break this down into a dozen distinct file-specific pull requests if that is preferred; the suggestion in the issue is that a single pull large pull request was the way to go though.

The are no changes to clause 19 and 25, and I deliberately stayed out of clause 29 as that clause is deliberately written using C idioms.

jwakely commented:
Thanks. I think a single pull request makes more sense. I'll go through this asap and merge.

tkoeppe commented on source/containers.tex:
```diff
@@ -2888,18 +2889,18 @@
+    using const_pointer          = typename allocator_traits<Allocator>::const_pointer;
+    using reference              = value_type&;
+    using const_reference        = const value_type&;
+    using size_type              = @\impdefnc@; // see \ref{container.requirements}
```
We can get rid of the non-correcting forms here now that the alignment at the ragged edge doesn't matter anymore.

AlisdairM replied:
If I understand correctly, you are saying replace '\impdefnc' with '\impdef'?
If so, is that a general rule, or applies only to header synopses and other code snippets?

tkoeppe replied:
Correct. "nc" stands for "no (italic) correction", which affects the column alignment. You ought to be able to see the difference (look for how characters (fail to) align on the grid without the correction). But now that the relevant text is at the end of the line, there's nothing left to be aligned, and we don't need to control the italic correction.

That was one of the benefits of changing `typedef` to `using` that @zygoloid pointed out in his first email about this subject a few months ago if I recall correctly.

tkoeppe commented on source/locales.tex:
```diff
@@ -4689,9 +4689,9 @@
-    typedef charT               char_type;
-    typedef InputIterator       iter_type;
-    typedef basic_string<charT> string_type;
+    using   char_type = charT;
```
Why right-align? Isn't that a bit cute? We used to left-align everything previously.

AlisdairM replied:
It is a bit cute. I largely threw it in to see the reaction, as we work out how to align the new syntax.  I like it, but suspected it might not be the preferred choice.  Unless I hear support, I will left-align any remaining right-aligned identifiers, but retain the alignment on the '='.

sigfpe replied:
Left-align is good.
-------- Original message --------From: Alisdair Meredith <notifications@github.com> Date: 4/16/2016  9:34 AM  (GMT-08:00) To: cplusplus/draft <draft@noreply.github.com> Subject: Re: [cplusplus/draft] [18-30] Replace typedefs with alias (#704) 
In source/locales.tex:
> @@ -4689,9 +4689,9 @@
>      class InputIterator = istreambuf_iterator<charT> >
>    class money_get : public locale::facet {
>    public:
> -    typedef charT               char_type;
> -    typedef InputIterator       iter_type;
> -    typedef basic_string<charT> string_type;
> +    using   char_type = charT;

It is a bit cute. I largely threw it in to see the reaction, as we work out how to align the new syntax.  I like it, but suspected it might not be the preferred choice.  Unless I hear support, I will left-align any remaining right-aligned identifiers, but retain the alignment on the '='.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub

tkoeppe commented on source/containers.tex:
```diff
@@ -2639,17 +2639,18 @@
+    using const_reference        = const T&;
+    using size_type              = size_t;
+    using difference_type        = ptrdiff_t;
+    using iterator               = @\impdefx{type of \tcode{array::iterator}}\itcorr[-1]@;
```
No more need for the italic correction at the end of the line now!

AlisdairM replied:
I'll admit, I am mostly moving text around and don't recognize all of the escape codes ;)
Will update, thanks.

AlisdairM commented:
Final (I hope!) patch adds the updates for <utility>.  I will now wait for this pull request to land before once-again rebasing https://github.com/cplusplus/draft/pull/683

tkoeppe commented:
Can you squash the few fixup commits you have into the respective main commits?

AlisdairM commented:
Squash them *all* as a single commit, or just the last couple?

tkoeppe commented:
Just the ones that are fixups - 0f6d327 and 8933e27.

lv-zheng commented:
typo in `source/support.tex` line 132: `using` is misspelt as `ussing`

AlisdairM commented:
Fix the typo (thanks lv-zheng) and force-pushed a single squash-commit of the whole branch.  No other changes.

zygoloid commented:
I want to go ahead with this, but not until after Oulu. Sorry if this requires extra rebasing as a result!

tkoeppe commented:
@zygoloid: Last standard, you didn't want to make major formatting changes after the CD went out (e.g. I remember fixing up the vertical whitespace problems with the old heading styles). Will there still be room for changes like this after Oulu? Will you only create the CD after the meeting?

AlisdairM commented:
If we plan to land this after Oulu, is there anything I can do to land it /before/ papers start landing from the meeting.  It will be much easier to update the standard for incoming papers with this in place (and editorial discretion when applying updates) than taking another pass to totally redo the work, at an undetermined time after all Oulu papers have landed again.

zygoloid commented:
If we're going to land this before we land the Oulu motions, we may as well do so now. I think you're right, handling the merge as Oulu motions land is not going to be very difficult. (Sorry for not merging this sooner, I'd missed that the pre-Oulu mailing deadline was 2pm not end-of-day, and was putting together the new working draft in a bit of a rush.)

zygoloid commented on source/containers.tex:
```diff
@@ -8345,11 +8346,11 @@
+    using reference       = typename Container::reference;
+    using const_reference = typename Container::const_reference;
+    using size_type       = typename Container::size_type;
+    using container_type  =          Container;
```
Maybe remove the whitespace before `Container` here. Now that we're not aligning anything else after the `=`, this looks a bit odd.

AlisdairM replied:
I like my 2-dimensional code :)  But defer to the actual project editor on style choice - will update this tonight.

zygoloid commented on source/future.tex:
```diff
@@ -1271,7 +1271,7 @@
 that take one or two arguments,
 many of the function objects in this standard
-correspondingly provide typedefs
+correspondingly provide type aliases
```
Maybe `\grammarterm{typedef-name}{s}`, to be clear that we don't care whether they're introduced by a `typedef` declaration or by an *alias-declaration*? "Type alias" is not a defined term in the standard.

AlisdairM replied:
Thanks - I think I may want to apply this recommendation in several places that I did not touch, as I did not have a better way to update the existing wording talking about typedefs.  I'll take another pass tonight.

AlisdairM commented:
Corrected Richard's issues above, and deployed the \grammarterm{typedef-name}{s} in several other places.  Tried to avoid reformatting around long lines, but at least one paragraph was re-wrapped.

Rebased, squashed, and --force pushed again.

zygoloid commented on source/iterators.tex:
```diff
@@ -3289,15 +3294,15 @@
+    using traits_type       = traits;
+    using streambuf_type    = basic_streambuf<charT,traits>;
+    using ostream_type = basic_ostream<charT,traits>;
 
```
Please align this.

AlisdairM replied:
Rebased again (just to be safe) and aligned the line highlighted above - should land cleanly.
